### PR TITLE
Switch to SQLite implementation of EntitiesRepository

### PR DESCRIPTION
**Pull request by seadowg**
Work towards #6012
~~Blocked by #6276~~

The major changes here:
- Entities can now be reset from the project management settings
- Local entities are now enabled by default
- Local entities created before an update/follow-up form is downloaded will be available in forms downloaded later
- Collect should support 100K item entity lists on most devices. For example with my Nexus 5X (which can barely handle the Play Store at this point):
    - Downloading form: ~2 mins (crashes after 30s when downloading manually however)
    - Loading form (with cached def): ~40s
    - Filtering entities by properties: ~2s

#### Why is this the best possible solution? Were any other approaches considered?

I think there are still a bunch of things we could do to get even better performance (both speed and memory wise) and to even make 1M item entity lists more credible, but this felt like a big enough milestone to merge before going any further.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

The big risk here is to filling and downloading entity list forms as it's mostly changes to how local entities are stored.

#### Before submitting this PR, please make sure you have:
- [x] added or modified tests for any new or changed behavior
- [x] run `./gradlew connectedAndroidTest` (or `./gradlew testLab`) and confirmed all checks still pass
- [x] added a comment above any new strings describing it for translators
- [x] added any new strings with date formatting to `DateFormatsTest`
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

### CODE REVIEW

**Comment by seadowg** [comment]
@getodk/testers this switches the implementation of entities storage over to a database. The goal here is to enable fast filtering for `name=` and `<property>=` filters on nodesets (like in a select) while also keeping memory use low (to avoid OOM errors). I think it'd be good to do an initial pass of the changes here with the performance testing project @lognaturel put together and get some first impressions of how we're doing. Here are some notes on the current performance expectations I'm seeing (so you can see if you get similar results):
- Downloading an entity update or follow-up form will be slightly (a few seconds) slower than downloading the same form in an older (pre-entity) version of Collect and memory use will spike during (so there could be a risk of OOMs here). Downloading one of these forms for a subsequent time will actually be slightly slower again, but not a lot.
- Loading entity update and follow-up forms will also be slightly slower than loading the same form in an older (pre-entity) version of Collect, but memory use will be much lower and will reduce after loading.
- Querying for entity properties/name on an entity should be about as fast as doing so with the same forms in an older (pre-entity) version of Collect and memory use shouldn't increase by much at all when moving between these screens.

I got to this kind of performance testing against the "100k Entities Filter" form with an emulator (running on a MacBook) so it could be that the results you see on a device are quite different. I'll have a play with an older phone as well, but thought it'd be good to get us all looking at it! What's for certain is I'm definitely seeing better overall performance (faster and lower memory use) compared than using these forms with what's currently on `master`.

EDIT: You'll also be able to clear entities from the standard project reset settings now.

**Comment by dbemke** [comment]
We tried 100k entities on Android 10 (Redmi devices) and it seems to be working better. On 2 devices with Android 10 no probelms on downloading the form, refreshing the list of blank forms or filling the form. On Android 14 it works well, too.

**Comment by seadowg** [comment]
> We tried 100k entities on Android 10 (Redmi devices) and it seems to be working better. On 2 devices with Android 10 no probelms on downloading the form, refreshing the list of blank forms or filling the form. On Android 14 it works well, too.

Great thanks! I think that's probably good enough for us to go ahead with reviewing/merging this, and then we can start more in depth performance testing/improvements after.

File: collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+
+private object ListsTable {
+    const val TABLE_NAME = "lists"
+    const val COLUMN_NAME = "name"
+}
+
+private object EntitiesTable {
+    const val COLUMN_ID = "id"
+    const val COLUMN_LABEL = "label"
+    const val COLUMN_VERSION = "vesion"
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt]: Typo in `vesion`.

File: collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+            CREATE TABLE IF NOT EXISTS ${ListsTable.TABLE_NAME} (
+                    $_ID integer PRIMARY KEY, 
+                    ${ListsTable.COLUMN_NAME} text NOT NULL
+            );
+            """.trimIndent()
+        )
+    }
+
+    override fun onUpgrade(db: SQLiteDatabase?, oldVersion: Int) {
+        TODO("Not yet implemented")
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt]: I think you can remove this `TODO` and the one below. This is not something to do for now and we might even never need to write any code here (assuming the db will never be changed).

File: collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+    }
+
+    override fun onDowngrade(db: SQLiteDatabase?) {
+        TODO("Not yet implemented")
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt]: In other DatabaseMigrators in `onDowngrade` we remove the whole table and let it be recreated. Generally downgrading should not be possible - you need to uninstall the newer version first but it might be possible using ADB... am I right? We should either do the same here (remove and recreate the table) or remove `onDowngrade` from  `DatabaseMigrator` at all. What do you think?

File: collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+                    list,
+                    cursor,
+                    cursor.getInt("rowid")
+                )
+            }
+    }
+
+    override fun clear() {
+        getLists().forEach {
+            databaseConnection.writeableDatabase.delete(it)
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt]: If I remove this all tests still pass.

File: collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+    private fun updateProperties(entity: Entity) {
+        entity.properties.map { it.first }.forEach {
+            try {
+                databaseConnection.writeableDatabase.execSQL(
+                    """
+                    ALTER TABLE ${entity.list} ADD $it text;
+                    """.trimIndent()
+                )
+            } catch (e: SQLiteException) {
+                println(e)
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt]: Should this be `Timber.e`?

File: collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+
+            execSQL(
+                """
+                CREATE UNIQUE INDEX ${list}_unique_id_index ON $list (${EntitiesTable.COLUMN_ID});
+                """.trimIndent()
+            )
+        }
+    }
+
+    private fun updateProperties(entity: Entity) {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt]: It does not update but just adds properties right? If so the name of this method could be improved.

File: collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+        getLists().forEach {
+            databaseConnection.writeableDatabase.execSQL(
+                """
+                DROP TABLE IF EXISTS ${it}_row_numbers;
+                """.trimIndent()
+            )
+
+            databaseConnection.writeableDatabase.execSQL(
+                """
+                CREATE TABLE ${it}_row_numbers AS SELECT _id FROM $it;
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt]: What is going on here? Why do you need that `_row_numbers` table if it only stores the ids?

**Review by grzesiek2010 — CHANGES_REQUESTED** [review]

File: collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt (hunk 7 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt]: This is there for the queries in `#getEntities`/`#getById`/`#getAllByProperty`. I think the best way to understand what's going on is to think about the `InMem...` implementation of `#getAllByProperty` which basically does:

```kotlin
entities.filter { it.list == list }.mapIndexed { index, entity ->
    Entity.Saved(
        entity.list,
        entity.id,
        entity.label,
        entity.version,
        entity.properties,
        entity.state,
        index,
        entity.trunkVersion,
        entity.branchId
    )
}.filter { entity ->
    entity.properties.any { (first, second) -> first == property && second == value
}
```

Here we're using the position in the list as the `index` for each entity and then filtering. In SQL that's a little more tricky as there's no simple way to get the "position" of a row in a table. Newer versions of SQLite actually have a `row_number()` function to handle this, but we can't use that on the versions of Android we support. One solution is to use a nested query like I do in [this commit](https://github.com/getodk/collect/pull/6290/commits/ea61916daf1e33f6b4b9bda7e45fc41c1a69a268), but that was very slow - SQLite probably ends up with something close to an "n squared" query to execute that. I found that creating an "index" table that maps IDs to row numbers was really fast to query and to recreate on any `#save` so I've ended up with this.

The core need here (that the tests enforce) is that `index`/`id` is consistent between queries (until the data changes) and that there are no gaps in `index` so that it works for multiplicity (how XPath will address the entities). Definitely let me know if you can think of other solutions though! It's a frustratingly tricky one.

File: collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt (hunk 3 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt]: Could you file a separate issue for this? It's probably worth us having a quick think about how downgrades work for the other DBs as well.

File: collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt (hunk 7 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt]: Ok now I understand but it definitely took me too much time to wrap my head around this concept. Please document the purpose of those `_row_numbers` tables in detail. It's also worth adding that once our min SDK version is bumped (API 29?) we will be able to use the `row_number()` function.

File: collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+            properties,
+            Entity.State.fromId(cursor.getInt(EntitiesTable.COLUMN_STATE)),
+            rowId - 1,
+            cursor.getIntOrNull(EntitiesTable.COLUMN_TRUNK_VERSION),
+            cursor.getString(EntitiesTable.COLUMN_BRANCH_ID)
+        )
+    }
+}
+
+class EntitiesDatabaseMigrator :
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt]: This can be private.

File: collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+        val propertyColumns = cursor.columnNames.filter {
+            !listOf(
+                _ID,
+                EntitiesTable.COLUMN_ID,
+                EntitiesTable.COLUMN_LABEL,
+                EntitiesTable.COLUMN_VERSION,
+                EntitiesTable.COLUMN_TRUNK_VERSION,
+                EntitiesTable.COLUMN_BRANCH_ID,
+                EntitiesTable.COLUMN_STATE,
+                "rowid"
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt]: Can we have a const for this column name as well? It might be worth adding a short comment that it's a column provided by sqlite.

File: collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+    private fun updatePropertyColumns(entity: Entity) {
+        entity.properties.map { it.first }.forEach {
+            try {
+                databaseConnection.writeableDatabase.execSQL(
+                    """
+                    ALTER TABLE ${entity.list} ADD $it text;
+                    """.trimIndent()
+                )
+            } catch (e: SQLiteException) {
+                // Ignored
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt]: I thought this catch clause was added to avoid crashes caused by some invalid column names or something like that. However, now I see that there is an exception when I run `#save updates existing properties` test. There is nothing unexpected/invalid there, right? I don't think we should try to avoid adding columns that already exist by catching that exception. Can't we check if a column with a given name already exists?

File: collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt [hunk 11]
```diff
@@ -0,0 +1,355 @@
...
+private object EntitiesTable {
+    const val COLUMN_ID = "id"
+    const val COLUMN_LABEL = "label"
+    const val COLUMN_VERSION = "version"
+    const val COLUMN_TRUNK_VERSION = "trunk_version"
+    const val COLUMN_BRANCH_ID = "branch_id"
+    const val COLUMN_STATE = "state"
+}
+
+class DatabaseEntitiesRepository(context: Context, dbPath: String) : EntitiesRepository {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt]: Shouldn't it be on the same level as other database repositories? The ones for forms, instances, savepoints are in the databases `package`.

File: androidshared/src/main/java/org/odk/collect/androidshared/sqlite/DatabaseConnection.kt [hunk 12]
```diff
@@ -1,4 +1,4 @@
-package org.odk.collect.android.database
+package org.odk.collect.androidshared.sqlite
```
Comment (grzesiek2010) [inline androidshared/src/main/java/org/odk/collect/androidshared/sqlite/DatabaseConnection.kt:1]: Don't you think it would make sense to move that sqlite package to a separate module? It might be for a separate pr of course.

File: androidshared/src/main/java/org/odk/collect/androidshared/sqlite/CursorExt.kt [hunk 13]
```diff
@@ -0,0 +1,70 @@
+package org.odk.collect.androidshared.sqlite
+
+import android.database.Cursor
+import androidx.core.database.getIntOrNull
+import androidx.core.database.getStringOrNull
+
+object CursorExt {
```
Comment (grzesiek2010) [inline androidshared/src/main/java/org/odk/collect/androidshared/sqlite/CursorExt.kt:7]: What's the point of having this object? I think we usually create extension functions on a file level without any objects like here.

File: androidshared/src/main/java/org/odk/collect/androidshared/sqlite/SQLiteDatabaseExt.kt [hunk 14]
```diff
@@ -0,0 +1,23 @@
+package org.odk.collect.androidshared.sqlite
+
+import android.database.Cursor
+import android.database.sqlite.SQLiteDatabase
+import android.database.sqlite.SQLiteQueryBuilder
+
+object SQLiteDatabaseExt {
```
Comment (grzesiek2010) [inline androidshared/src/main/java/org/odk/collect/androidshared/sqlite/SQLiteDatabaseExt.kt:7]: What's the point of having this object? I think we usually create extension functions on a file level without any objects like here.

File: androidshared/src/main/java/org/odk/collect/androidshared/sqlite/DatabaseMigrator.java [hunk 15]
```diff
@@ -1,4 +1,4 @@
-package org.odk.collect.android.database;
+package org.odk.collect.androidshared.sqlite;
```
Comment (grzesiek2010) [inline androidshared/src/main/java/org/odk/collect/androidshared/sqlite/DatabaseMigrator.java:1]: It would be good to move `SQLiteUtils` class here too and probably refactor (add extensions functions instead of static ones).

File: androidshared/src/main/java/org/odk/collect/androidshared/sqlite/CursorExt.kt [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+
+            accumulator
+        }
+    }
+
+    fun <T> Cursor.first(map: (Cursor) -> T): T? {
+        return this.use {
+            if (it.count >= 1) {
+                it.moveToPosition(-1)
+                it.moveToNext()
```
Comment (grzesiek2010) [inline androidshared/src/main/java/org/odk/collect/androidshared/sqlite/CursorExt.kt]: You can simplify this:

```
return this.use {
            if (it.moveToFirst()) {
                map(it)
            } else {
                return null
            }
        }
```

File: entities/src/main/java/org/odk/collect/entities/storage/Entity.kt [hunk 17]
```diff
@@ -35,9 +35,15 @@ sealed interface Entity {
         override val branchId: String = UUID.randomUUID().toString()
     ) : Entity
 
-    enum class State {
-        OFFLINE,
-        ONLINE
+    enum class State(val id: Int) {
+        OFFLINE(0),
```
Comment (grzesiek2010) [inline entities/src/main/java/org/odk/collect/entities/storage/Entity.kt]: Wouldn't it be easier to just use the string representation instead of numbers? If you call `toString`  on `State.OFFLINE` you should get `OFFLINE` (the same with `State.ONLINE`). You can store such a string in the database.

File: entities/src/main/java/org/odk/collect/entities/LocalEntityUseCases.kt [hunk 18]
```diff
@@ -54,38 +55,43 @@ object LocalEntityUseCases {
         serverList: File,
         entitiesRepository: EntitiesRepository
     ) {
+        Timber.d("SQLite: starting local entity update")
```
Comment (grzesiek2010) [inline entities/src/main/java/org/odk/collect/entities/LocalEntityUseCases.kt]: Do we need those logs?

File: collect_app/src/androidTest/java/org/odk/collect/android/feature/settings/ResetProjectTest.kt [hunk 19]
```diff
@@ -3,67 +3,76 @@ package org.odk.collect.android.feature.settings
...
-            .assertTextDoesNotExist("All widgets")
+
+            .startBlankForm("One Question Entity Update")
+            .assertTextDoesNotExist("Logan Roy")
     }
 
     @Test
-    fun adminSettings_shouldBeReset() {
-        rule.startAtMainMenu()
+    fun canResetAdminSettings() {
```
Comment (grzesiek2010) [inline collect_app/src/androidTest/java/org/odk/collect/android/feature/settings/ResetProjectTest.kt]: I think the terminology we have been using is `unprotected`/`protected` instead of `user` (or whatever)/`admin` settings.

File: collect_app/src/androidTest/java/org/odk/collect/android/feature/settings/ResetProjectTest.kt [hunk 20]
```diff
@@ -127,4 +137,17 @@ class ResetApplicationTest {
...
     }
+
+    @Test
+    fun when_rotateScreen_should_resetDialogNotDisappear() {
+        rule.startAtFirstLaunch()
+            .clickTryCollect()
+            .openProjectSettingsDialog()
+            .clickSettings()
+            .clickProjectManagement()
+            .clickOnResetApplication()
```
Comment (grzesiek2010) [inline collect_app/src/androidTest/java/org/odk/collect/android/feature/settings/ResetProjectTest.kt]: Please rename this method to `resetProject` to make it consistent with the class name that you renamed.

**Review by grzesiek2010 — CHANGES_REQUESTED** [review]

File: collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt (hunk 7 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt]: >  It's also worth adding that once our min SDK version is bumped (API 29?) we will be able to use the row_number() function.

I wanted to initially focus on experimenting with solutions that work on every device, but I'm also now curious about the performance we'd actually get with `row_number()`. I'll add an issue to investigate later.

File: collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt (hunk 10 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt]: Annoyingly SQLite does not support a statement like `ADD COLUMN IF NOT EXISTS`, but we can inspect the table schema using `PRAGMA` (https://www.sqlite.org/pragma.html#pragma_table_info). I did look into that briefly, but it was definitely a little nastier and also requires a read query whereas this solution can just try and fail.

File: androidshared/src/main/java/org/odk/collect/androidshared/sqlite/DatabaseConnection.kt (hunk 12 above)
Comment (seadowg) [reply androidshared/src/main/java/org/odk/collect/androidshared/sqlite/DatabaseConnection.kt:1]: Absolutely! Probably worth doing as a follow-up at some point though.

File: androidshared/src/main/java/org/odk/collect/androidshared/sqlite/CursorExt.kt (hunk 13 above)
Comment (seadowg) [reply androidshared/src/main/java/org/odk/collect/androidshared/sqlite/CursorExt.kt:7]: I guess I prefer them in an `object` as it makes them addressable easily via Java (you don't need to use the generated `...Kt` class). I think it also feels natural to me for extensions to be grouped under something other than just the package - that seems a little more organized. What do you think?

File: androidshared/src/main/java/org/odk/collect/androidshared/sqlite/SQLiteDatabaseExt.kt (hunk 14 above)
Comment (seadowg) [reply androidshared/src/main/java/org/odk/collect/androidshared/sqlite/SQLiteDatabaseExt.kt:7]: https://github.com/getodk/collect/pull/6290#discussion_r1708904304

File: entities/src/main/java/org/odk/collect/entities/storage/Entity.kt (hunk 17 above)
Comment (seadowg) [reply entities/src/main/java/org/odk/collect/entities/storage/Entity.kt]: I think that would add quite a lot more data though as you'd end up with "OFFLINE" or "ONLINE" in every entity row (which we're hoping we can get into the millions).

File: androidshared/src/main/java/org/odk/collect/androidshared/sqlite/DatabaseMigrator.java (hunk 15 above)
Comment (seadowg) [reply androidshared/src/main/java/org/odk/collect/androidshared/sqlite/DatabaseMigrator.java:1]: I've moved it (and the classes it depends on), but I'll leave the refactoring. I'll create an issue to create a module for the database code and to consolidate `SQLiteDatabaseExt` and `SQLiteUtils`.

Comment (grzesiek2010) [reply androidshared/src/main/java/org/odk/collect/androidshared/sqlite/DatabaseMigrator.java:1]: Yeah, we can leave the refactoring but please move the test class (`SQLiteUtilsTest`) too.

File: collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt (hunk 10 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt]: What about the `#doesColumnExist` from `SQLiteUtils`? Can't we add an if statement before trying to call `ALTER TABLE ${entity.list} ADD $it text;`?

File: androidshared/src/main/java/org/odk/collect/androidshared/sqlite/CursorExt.kt (hunk 13 above)
Comment (grzesiek2010) [reply androidshared/src/main/java/org/odk/collect/androidshared/sqlite/CursorExt.kt:7]: > What do you think?

I think I'm used to the approach I've described so that's probably why I like it more but no strong preferences here.

File: entities/build.gradle.kts [hunk 21]
```diff
@@ -64,6 +64,7 @@ dependencies {
     implementation(Dependencies.android_material)
     implementation(Dependencies.androidx_navigation_fragment_ktx)
     implementation(Dependencies.androidx_navigation_ui)
+    implementation(Dependencies.timber)
```
Comment (grzesiek2010) [inline entities/build.gradle.kts]: I think this is not needed anymore.

File: entities/src/main/java/org/odk/collect/entities/storage/Entity.kt (hunk 17 above)
Comment (grzesiek2010) [reply entities/src/main/java/org/odk/collect/entities/storage/Entity.kt]: Oh interesting. Could you add a comment explaining that? I'm pretty sure I would be tempted to refactor that code next time I work with it.

**Review by grzesiek2010 — CHANGES_REQUESTED** [review]

File: collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt (hunk 3 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt]: https://github.com/getodk/collect/issues/6315

File: collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt (hunk 10 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt]: `#doesColumnExist` does a `SELECT *...` and then grabs the column names which could be really slow. I'm just working on a benchmark test (#6318) that'll make it a lot easier to play around with the effects of different implementations, so I'd rather wait for that to make changes that might change performance if that's OK?

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt]: > What about the #doesColumnExist from SQLiteUtils?

Ah good shout! I hadn't spotted it before. It looks like its implementation could be improved a little (it currently does a full `SELECT *...` and then just grabs the columns). I can use that for the moment, and if we find we need a speed-up we can switch it to use the `PRAGMA` stuff or fall back to ignoring the exception.

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt]: Actually can we come back to this? I'm having a hard time getting `#doesColumnExist` working within the transaction (I'm not sure why), and I don't really want to get hung up on it for the moment if everything is currently working.

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt]: Ok, let's leave it for now.

File: entities/src/main/java/org/odk/collect/entities/storage/Entity.kt (hunk 17 above)
Comment (seadowg) [reply entities/src/main/java/org/odk/collect/entities/storage/Entity.kt]: Sure! I'll also move the conversion to `DatabaseEntitiesRepository` as `Entity` shouldn't really need to know anything about "serializing" the enum.

**Review by grzesiek2010 — APPROVED** [review]

**Comment by WKobus** [comment]
**Tested with Success**

Verified on device with Android 14

**Verified cases:**
- Resetting entities in project management
- Performance on 100k entities form
- Downloading update form after creating local entity
- Regression check on entities

**Comment by dbemke** [comment]
**Tested with Success**

Verified on a device with Android 10 and Android 8.1
